### ISSUE TITLE
Plans to support `const`?

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#179]
`const` is part of ES6, which has apparently reached Release Candidate status.  How difficult would it be to implement const support, and what would the timeframe be on such an implementation?

The reason I ask is because old versions of SpiderMonkey (and possibly current ones, I'm not totally sure about that though) recognized `const` and aliased it to `var`.  As a result, Sphere, the game engine of which my minisphere is a clone of, had many games made for it that included `const` declarations and are broken under minisphere as a result, since Duktape doesn't support it and throws a syntax error.

**Comment by svaarala** [comment]
I'll need to write an internal document about the gap to ES6 and what needs to be done - it's a rather large set of work because some of the features require compiler changes, data structure changes, garbage collector changes, etc. So I've been adding specific features from ES6 so far, but there's going to be a limit at which point larger rewrites will be needed. For example, multiple return values would be very useful but require significant compiler rework. Yet another consideration is footprint: full ES6 support will significantly increase footprint, so one important feature will be the ability to more flexibly drop undesired built-ins (currently you can disable features but the set of built-in objects remains the same).

Regarding `const`, it'd be trivial to simply treat it like `var`. More proper support needs write protection, and it'd also be useful to assign constant numbers and strings to the bytecode constant table (instead of binding them unnecessarily to a register).

So in short, it's certainly possible and shouldn't be too difficult.

**Comment by JoshEngebretson** [comment]
Ok, as you mentioned ES6 in general :)  Multiline strings may be low lying fruit and would be super handy

**Comment by svaarala** [comment]
@JoshEngebretson Agree with that :)

**Comment by fatcerberus** [comment]
Would it be feasible to implement `const` within 1.4.0 timeframe?  This is the main sticking point right now preventing full compatibility of minisphere with the old Sphere engine.

**Comment by svaarala** [comment]
A minimal hack would be just to treat `const` as a synonym to `var` for the time being. Obviously wrong in many ways, but would make most code work (after all, how much real code tries to write constants *and* relies on that working correctly, i.e. to be rejected?). This would be just a few hours work (testcases etc are the main part).

But going forward, a minimal actual implementation would be something like:

- Implement `const` as a normal variable. They would then be register mapped *and* writable.
- Preventing writes to them would require compiler changes to reject a write to an identifier bound to a register.
- Slow path code would need to know both that the identifier was register mapped and that it was a constant, so that writes are properly rejected.
- Closing an environment record would need a change to ensure constants remain unwritable once the environment record closes. As the value are copied from registers, additional metadata would be needed to indicate that the register was a const.

Alternatively `const` variables could be mapped to the const table:

- Add necessary metadata to know that an identifier is in the const table and not a register.
- Add support for compile time identifier reference ending up constant bound (now it's either register bound or slow path).
- Slow path changes to use additional metadata for protecting constants.
- Environment record closing would need to support this too so that certain identifiers are copied from the constant table with property attributes preventing a write. Some additional metadata would be needed for this.

I'm not sure of this is doable within 1.4. I will need to touch the environment closing process as part of #229, so maybe there's some synergy there.

**Comment by fatcerberus** [comment]
Synonym for `var` would be an acceptable hack for now, in fact that's exactly how the ancient SpiderMonkey in Sphere 1.5 treats it I believe.  But of course in the longer term a correct implementation would be better. :)

**Comment by svaarala** [comment]
That should be doable for 1.4 but I'd prefer it to be disabled by default because it really is a hack :)

**Comment by fatcerberus** [comment]
I was actually going to do this in my own copy of Duktape a while back, but I couldn't figure out how the lexer works.  For example, where are all the keywords?

**Comment by svaarala** [comment]
Keywords are supplied by `genstrings.py` and they are in 1:1 correspondence with token numbering so that a simple add/subtract is enough to convert a built-in string index into a token number.

**Comment by svaarala** [comment]
I'll post a quick hack branch soon for you to test.

**Comment by svaarala** [comment]
Here's a quick hack: #360.

**Comment by fatcerberus** [comment]
Does ES6 specifically *require* writes to a `const` to be rejected at compile time?  If not it would probably be easier to simply implement it similar to non-writable property: runtime error in strict mode, silent failure otherwise.

Knowing how ES6 turned out in general though I'm going to guess the compile-time check is mandatory. :-/

**Comment by svaarala** [comment]
It's listed as a future reserved word and they have special handling, but Section 2 states:

> A conforming implementation of ECMAScript is permitted to support program and regular expression syntax not described in this specification. In particular, a conforming implementation of ECMAScript is permitted to support program syntax that makes use of the “future reserved words” listed in 7.6.1.2 of this specification.

So I think it'd be OK to support it by default from compliance perspective.

This initial `const` placeholder would not be ES6 compliant and would undergo an incompatible change later - but I don't see that as a big issue.

**Comment by fatcerberus** [comment]
Strictly speaking, it would be incompatible--but as you say how many real world programs are going to declare something `const` and then expect changing the value to work properly?  We can just say it's "undefined behavior". ;-)

**Comment by svaarala** [comment]
I agree, I don't see it as a big issue. Some programs do rely on protection semantics (that's why there are property attributes too) and in sandboxing they can even be critical. But as long as it's clear this `const` is not protected against change (maybe a note in sandboxing document), it should be OK.

**Comment by svaarala** [comment]
#360 now adds minimal syntax support for `const`; const variables are treated like `var` declared variables, except that an initializer is required.

**Comment by fatcerberus** [comment]
Oh good, that's one less branch I have to octopus-merge when updating Duktape for minisphere.  Thanks! :)

**Comment by fatcerberus** [comment]
Considering the minimal `const` support in master, this can probably be closed.

**Comment by svaarala** [comment]
Sounds good - but opened #1148 to track its proper implementation.
